### PR TITLE
test: Display benchmark results with thousands separators

### LINKED ISSUE #10953: Benchmark times should not display all significant figures.
```rust
#[feature(macro_rules, asm)];

extern mod extra;
use BH = extra::test::BenchHarness;
macro_rules! mk_benches(
    ( $($name:ident, $n:expr;)* ) => {

        $(
            #[bench]
            fn $name(bh: &mut BH) {
                bh.iter(|| for _ in range(0, $n) { unsafe {asm!("")}; } );
            }
            )*
    })

mk_benches! {
    a,           1;
    b,          40; // to get a meaningful time
    c,         100;
    d,       1_000;
    e,      10_000;
    f,     100_000;
    g,   1_000_000;
    h,  10_000_000;
    i, 100_000_000;
}
```

Compiled with `rustc -O --test` (using a rustc with https://github.com/mozilla/rust/pull/10952) shows

```
running 9 tests
test a ... bench:         2 ns/iter (+/- 1)
test b ... bench:        34 ns/iter (+/- 3)
test c ... bench:        62 ns/iter (+/- 8)
test d ... bench:       383 ns/iter (+/- 60)
test e ... bench:      3608 ns/iter (+/- 367)
test f ... bench:     35984 ns/iter (+/- 4093)
test g ... bench:    361734 ns/iter (+/- 98714)
test h ... bench:   3381412 ns/iter (+/- 281941)
test i ... bench:  34584793 ns/iter (+/- 2991062)
```

The long numbers are hard to read, and most of the digits are unnecessary. It would be clearer if we only showed a fixed precision, e.g., using [engineering notation](http://en.wikipedia.org/wiki/Engineering_notation):

```
running 9 tests
test a ... bench:         2 ns/iter (+/-     1)
test b ... bench:        34 ns/iter (+/-     3)
test c ... bench:        62 ns/iter (+/-     8)
test d ... bench:       383 ns/iter (+/-    60)
test e ... bench:     3.6e3 ns/iter (+/-   367)
test f ... bench:    36.0e3 ns/iter (+/- 4.1e3)
test g ... bench:   361.7e3 ns/iter (+/-  99e3)
test h ... bench:     3.4e6 ns/iter (+/- 280e3)
test i ... bench:    34.6e6 ns/iter (+/- 3.0e6)
```

This specific notation has the advantages of easy local comparison, that is, tests that take approximately the same amount of time can easily be compared without accidentally missing `8.0e3` vs `1.0e4` (`8e3` vs `10e3` under the proposed scheme).

This could also be achieved by changing the units (i.e. `ns` -> `us` -> `ms`), but I personally think that that difference is more subtle than necessary, but I don't have a particularly strong opinion on it.

```
running 9 tests
test a ... bench:         2 ns/iter (+/-    1)
test b ... bench:        34 ns/iter (+/-    3)
test c ... bench:        62 ns/iter (+/-    8)
test d ... bench:       383 ns/iter (+/-   60)
test e ... bench:       3.6 us/iter (+/- 0.37)
test f ... bench:      36.0 us/iter (+/-  4.1)
test g ... bench:     361.7 us/iter (+/-   99)
test h ... bench:       3.4 ms/iter (+/- 0.28)
test i ... bench:      34.6 ms/iter (+/-  3.0)
```

### LINKED ISSUE #26109: bench should put deliminators on nanosecond numbers
From [this initial rust experience blog](http://ericsink.com/entries/rust1.html), `cargo bench` (I assume identical to `rustc --test`) should report nanosecond times as `3_000_000` or `3,000,000`, not `3000000`.

This reportedly would require fixing in libtest.

### PR DESCRIPTION
**Pull request by bluss**
test: Display benchmark results with thousands separators

Example display:

```
running 9 tests
test a ... bench:           0 ns/iter (+/- 0)
test b ... bench:          52 ns/iter (+/- 0)
test c ... bench:          88 ns/iter (+/- 0)
test d ... bench:         618 ns/iter (+/- 111)
test e ... bench:       5,933 ns/iter (+/- 87)
test f ... bench:      59,280 ns/iter (+/- 1,052)
test g ... bench:     588,672 ns/iter (+/- 3,381)
test h ... bench:   5,894,227 ns/iter (+/- 303,489)
test i ... bench:  59,112,382 ns/iter (+/- 1,500,110)
```

Fixes #10953
Fixes #26109

### CODE REVIEW

**Comment by bluss** [comment]
r? @huonw 

I think scientific notation is too hard to read. The nicest improvement is rounding to sig figs, units are a bit tricky to notice too.

**Comment by bluss** [comment]
Maybe four sigfigs & 1 sigfig deviation is better

example

```
running 9 tests
test a ... bench:         0 ns/iter (+/- 0)
test b ... bench:        52 ns/iter (+/- 0)
test c ... bench:        87 ns/iter (+/- 0)
test d ... bench:       617 ns/iter (+/- 1)
test e ... bench:     5.913 us/iter (+/- 0.02)
test f ... bench:     58.87 us/iter (+/- 0.03)
test g ... bench:     588.7 us/iter (+/- 0.7)
test h ... bench:     5.885 ms/iter (+/- 0.03)
test i ... bench:     58.88 ms/iter (+/- 0.2)
```

Example from real benchmarks

```
running 10 tests
test slice_iter              ... bench:        24 ns/iter (+/- 2)
test slice_iter_rev          ... bench:        24 ns/iter (+/- 0)
test stride_iter             ... bench:        24 ns/iter (+/- 1)
test stride_iter_rev         ... bench:        24 ns/iter (+/- 0)
test zip_loop                ... bench:       923 ns/iter (+/- 53)
test zip_loop3               ... bench:     1.372 us/iter (+/- 0.1)
test zip_slices_default_zip  ... bench:       935 ns/iter (+/- 9)
test zip_slices_default_zip3 ... bench:     1.383 us/iter (+/- 0.08)
test zip_slices_ziptuple     ... bench:       941 ns/iter (+/- 40)
test zipslices               ... bench:       942 ns/iter (+/- 167)

test result: ok. 0 passed; 0 failed; 0 ignored; 10 measured
```

**Comment by huonw** [comment]
Hm, as you say, the distinction between units is very subtle, especially the two most common unfortunately (n vs. u).

**Comment by Gankro** [comment]
More than anything it seems harder to "follow the trend" visually. Actual
charts and diagrams seems like a richer way to handle this sort of thing.

**Comment by bluss** [comment]
Let's do thousands separators instead?

**Comment by huonw** [comment]
That might be better yeah.

**Comment by bluss** [comment]
They look like this

```
running 9 tests
test a ... bench:           0 ns/iter (+/- 0)
test b ... bench:          52 ns/iter (+/- 0)
test c ... bench:          88 ns/iter (+/- 0)
test d ... bench:         618 ns/iter (+/- 111)
test e ... bench:       5,933 ns/iter (+/- 87)
test f ... bench:      59,280 ns/iter (+/- 1,052)
test g ... bench:     588,672 ns/iter (+/- 3,381)
test h ... bench:   5,894,227 ns/iter (+/- 303,489)
test i ... bench:  59,112,382 ns/iter (+/- 1,500,110)

test result: ok. 0 passed; 0 failed; 0 ignored; 9 measured
```

I think that's good enough. We could round it, but more correct was already harder to read.

**Comment by bluss** [comment]
Updated branch. I guess the thousands formatting was pretty quick and dirty.

File: src/libtest/lib.rs [hunk 1]
```diff
@@ -625,17 +625,44 @@ impl<T: Write> ConsoleTestState<T> {
...
+// Format a number with thousands separators
+fn fmt_thousands_sep(mut n: usize, sep: char) -> String {
+    use std::fmt::Write;
+    let mut output = String::new();
+    let mut first = true;
+    for &pow in &[9, 6, 3, 0] {
+        let base = 10_usize.pow(pow);
+        if pow == 0 || n / base != 0 {
+            if first {
+                output.write_fmt(format_args!("{}", n / base)).ok();
```
Comment (huonw) [inline src/libtest/lib.rs]: `ok` -> `unwrap()`?

Comment (bluss) [inline src/libtest/lib.rs]: I guess that's more idiomatic. On a string it doesn't matter.

**Comment by bluss** [comment]
Updated to use .unwrap().

**Comment by huonw** [comment]
@bors r+

(Now there's just the question of localising it properly...)

**Pull request merged**